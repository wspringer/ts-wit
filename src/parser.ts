import { grammar } from "./grammar";
import { defineModel } from "./semantics";
import { Wit } from "./ast";

export function parseWit(input: string): Wit {
  const semantics = grammar.createSemantics();
  defineModel(semantics);
  const match = grammar.match(input);
  if (match.failed()) {
    throw new Error(match.message);
  }
  return semantics(match).toModel();
}
