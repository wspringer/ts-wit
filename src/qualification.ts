import {
  Func,
  InterfaceDef,
  MetaOf,
  NestedPackage,
  TypeDef,
  Wit,
  World,
} from "./ast";
import { isInterfaceDef } from "./guards";

export type QualificationProps = {
  fqn: string;
};

export type Qualified = {
  TypeDef: QualificationProps;
  InterfaceDef: QualificationProps;
  NestedPackage: QualificationProps;
  Func: QualificationProps;
  World: QualificationProps;
};

export function qualify(wit: Wit): Wit<Qualified> {
  return {
    ...wit,
    interfaces: wit.interfaces.map((iface) =>
      qualifyInterfaceDef(wit.name || "", iface)
    ),
    worlds: wit.worlds.map((world) => qualifyWorld(wit.name || "", world)),
  } as Wit<Qualified>;
}

export function qualifyInterfaceDef(
  scope: string,
  iface: InterfaceDef
): InterfaceDef<Qualified> {
  const fqn = `${scope}/${iface.name}`;
  return {
    name: iface.name,
    functions: iface.functions.map((func) => qualifyFunc(fqn, func)),
    typeDefs: iface.typeDefs.map((typeDef) => qualifyTypeDef(fqn, typeDef)),
    fqn,
  };
}

export function qualifyTypeDef(
  scope: string,
  typeDef: TypeDef
): TypeDef<Qualified> {
  const fqn = `${scope}/${typeDef.name}`;
  return {
    ...typeDef,
    fqn,
  } as TypeDef<Qualified>;
}

export function qualifyNestedPackage(
  scope: string,
  pkg: NestedPackage
): NestedPackage<Qualified> {
  const fqn = `${scope}/${pkg.name}`;
  return {
    ...pkg,
    fqn,
    interfaces: pkg.interfaces.map((iface) => qualifyInterfaceDef(fqn, iface)),
  };
}

export function qualifyFunc(scope: string, func: Func): Func<Qualified> {
  return {
    ...func,
    fqn: `${scope}/${func.name}`,
  };
}

export function qualifyWorld(scope: string, world: World): World<Qualified> {
  const fqn = `${scope}/${world.name}`;
  return {
    ...world,
    typeDefs: world.typeDefs.map((typeDef) => qualifyTypeDef(fqn, typeDef)),
    exports: {
      functions: world.exports.functions.map((func) => qualifyFunc(fqn, func)),
      interfaces: world.exports.interfaces.map((iface) =>
        isInterfaceDef(iface) ? qualifyInterfaceDef(fqn, iface) : iface
      ),
    },
    imports: {
      functions: world.imports.functions.map((func) => qualifyFunc(fqn, func)),
      interfaces: world.exports.interfaces.map((iface) =>
        isInterfaceDef(iface) ? qualifyInterfaceDef(fqn, iface) : iface
      ),
    },
    fqn,
  } as World<Qualified>;
}
