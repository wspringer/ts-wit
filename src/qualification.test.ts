import { parseWit } from "./parser";
import { qualify } from "./qualification";
describe("qualification", () => {
  it("should qualify a Wit", () => {
    const wit = parseWit(`
      package foo:bar@1.0.0;

      world ranger {
        export sky: interface {
          type weird = string;
        }
        import limit: interface {
          resolve: func() -> result<string, string>;
        }
      }

      interface foo {
        type weird = string;
      }
    `);

    const qualified = qualify(wit);
    expect(qualified).toMatchSnapshot();
  });
});
