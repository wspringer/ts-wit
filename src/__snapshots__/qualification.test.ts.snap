// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`qualification should qualify a Wit 1`] = `
{
  "interfaces": [
    {
      "fqn": "foo:bar@1.0.0/foo",
      "functions": [],
      "name": "foo",
      "typeDefs": [
        {
          "fqn": "foo:bar@1.0.0/foo/weird",
          "kind": "alias",
          "name": "weird",
          "type": "string",
        },
      ],
    },
  ],
  "name": "foo:bar@1.0.0",
  "packages": [],
  "uses": [],
  "worlds": [
    {
      "exports": {
        "functions": [],
        "interfaces": [
          {
            "fqn": "foo:bar@1.0.0/ranger/sky",
            "functions": [],
            "name": "sky",
            "typeDefs": [
              {
                "fqn": "foo:bar@1.0.0/ranger/sky/weird",
                "kind": "alias",
                "name": "weird",
                "type": "string",
              },
            ],
          },
        ],
      },
      "fqn": "foo:bar@1.0.0/ranger",
      "imports": {
        "functions": [],
        "interfaces": [
          {
            "fqn": "foo:bar@1.0.0/ranger/sky",
            "functions": [],
            "name": "sky",
            "typeDefs": [
              {
                "fqn": "foo:bar@1.0.0/ranger/sky/weird",
                "kind": "alias",
                "name": "weird",
                "type": "string",
              },
            ],
          },
        ],
      },
      "includes": [],
      "name": "ranger",
      "typeDefs": [],
    },
  ],
}
`;
